Add support for certificate operations

jonandernovella opened the pull request:
- Individual certificates can be issued, updated, previewed and verified
- They are stored in a S3 backend
- Tests work with a mock S3 lib

jonandernovella committed:
Add deps to main container

jonandernovella committed:
Add support for issuing certificates

Certs are filled based on a pdf template and using pdfrw in order to write to the annotations.

jonandernovella committed:
Fix farfar key

jonandernovella committed:
Add preview, issue and update endpoints

jonandernovella committed:
Update db upon issue

jonandernovella committed:
Cleanup

jonandernovella committed:
Add more docstrings

jonandernovella committed:
Add static method for getting signer

jonandernovella committed:
Use static method in endpoint

jonandernovella committed:
Add first s3 components

jonandernovella committed:
Send proper status codes

jonandernovella committed:
Remove static method ref

jonandernovella committed:
Add more s3 stuff

jonandernovella committed:
More cleanup

jonandernovella committed:
Add method for checksum verification

jonandernovella committed:
Simplify s3 setup

jonandernovella committed:
Add first verification components

jonandernovella committed:
Use logical folder

jonandernovella committed:
Use same s3 path for getting checksum

jonandernovella committed:
Improve error handling

jonandernovella committed:
Define s3 vars in settings

jonandernovella committed:
Set cert settings in a better place

jonandernovella committed:
Use the defined settings

jonandernovella committed:
Embed hostname in qr code

jonandernovella committed:
Do not use HOSTNAME env var

jonandernovella committed:
Clean up to make linter happier

jonandernovella committed:
Make the linter happy

jonandernovella committed:
Fix import

jonandernovella committed:
Add more docstrings

jonandernovella committed:
Make endpoint accept PATCH

jonandernovella committed:
Handle valid signature but not present file

jonandernovella committed:
Change url in qr code

HenrikeW committed:
Add update button and pass props

adding an additional button in individual_view and passing props
to individual_certificate to decide what action the component should be used for.

HenrikeW committed:
Change button text depending on action

Button fpr the last step in individual_certificate should change text depending on if the user is going to update or issue a certificate.

HenrikeW committed:
fix merge conflicts in individual_edit

HenrikeW committed:
add update-cert-functionality to frontend

changes in individual_certificate:
- added updateCertificate function
- adjusted JSX to show update/issue specific things

jonandernovella committed:
Add missing dependencies to load script

jonandernovella committed:
Assign digital ids when issuing certs

The digital ids are present on the individual database for every entry, but they are only effective when assigned to the certificate attribute.

jonandernovella committed:
Create digital certificate id attribute

This id starts at 100000 and should be unique.

jonandernovella committed:
Use logger instead of print()

jonandernovella committed:
Log verification exception better

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

jonandernovella committed:
Log update exception better

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

jonandernovella committed:
Explicitely initialise boolean vars

jonandernovella committed:
Use request.json

HenrikeW committed:
adding content-type to preview-request

adding "Content-Type": "application/json" to the previewCertificate-
function fixes the bug in the preview. Now data that was inserted in the
form should does show up in the preview.

HenrikeW committed:
improving issueCertificate

jonandernovella committed:
Rework error handling for updating certs

HenrikeW committed:
improve error handling

HenrikeW committed:
removing console.log

HenrikeW committed:
cleaning code

HenrikeW committed:
Making error message in issueCertificate more specific

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

HenrikeW committed:
removing redundant code from error handling in updateCertificate

HenrikeW committed:
make ternary operator more specific / use switch instead

jonandernovella committed:
Move sequence creation to init

jonandernovella committed:
Allow nullable digital cert ids

jonandernovella committed:
Get post data uniformly

jonandernovella committed:
Allow get request for previews

jonandernovella committed:
Move get data to correct scope

jonandernovella committed:
Check for database type

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

HenrikeW committed:
Add button to to verify certificate

button added on individual_view, component it leads to does not exist yet

HenrikeW committed:
Create component for certificate verification

HenrikeW committed:
Create component for certificate verification

HenrikeW committed:
Basic layout for certificate verification

also started to write function for uploading. Still not functional

HenrikeW committed:
fetch-request sending arraybuffer

jonandernovella committed:
Add migration from 4 to 5

This includes the changes for digital certificate ids.

jonandernovella committed:
Bump schema version

jonandernovella committed:
Make the sequence optional

pontus committed:
Create simpler with sqlite

HenrikeW committed:
move fetch into reader.onload

This way the request is sent when clicking on the button for the first time.
(Before, you had to click two times --> bug)

HenrikeW committed:
add success box with styling

HenrikeW committed:
handle outdated certificate

show information to the user and fetch to get the latest cert from S3

HenrikeW committed:
handle invalid certificate

show error message to the user if uploaded file is not a valid certificate for the rabbit in question

jonandernovella committed:
Handle S3 related vars properly

jonandernovella committed:
Don't hardcode template file

jonandernovella committed:
Add tests for certificate related endpoints

S3 mocking is done using moto.

jonandernovella committed:
Set up certs for testing

HenrikeW committed:
Hide pdf when new file is chosen

introduce showpdf state variable and tidy up the code

jonandernovella committed:
Make sure we update the db upon cert updates

jonandernovella committed:
Improve tests to match expectations

jonandernovella committed:
Use consistent name for colour refs

HenrikeW committed:
change swedish text the user sees with an outdated cert

jonandernovella committed:
Test verify using unsigned certs

jonandernovella committed:
Make black happy

jonandernovella committed:
Remove quotation marks from env vars

HenrikeW committed:
take away error constructor in verification

HenrikeW committed:
Fixing the Error constructor

Replacing the Error icon with Warning icon to be able to use Error constructor.
Also restructuring the verifyCertificate function.

jonandernovella committed:
Try to make python linters happy

jonandernovella committed:
Add more fixes

jonandernovella committed:
Add more cleanup

pontus committed:
More linter happiness

pontus committed:
Fix some shellcheck complaints also

jonandernovella committed:
Fix import conflict

jonandernovella committed:
Remove host port forwarding for s3 container

pontus commented:
Test database shouldn't be checked in. Can we get a script to generate test keys/certificates rather than having them checked in?

pontus commented on frontend/src/individual_certificate.tsx:
```diff
@@ -370,26 +420,46 @@ export function IndividualCertificate({ id }: { id: string }) {
+          )}
+          <div className={style.paneControls}>
+            <a
+              target="_blank"
+              href={certificateUrl}
+              download={individual.number}
```
Do we really need `target="_blank"` here? I have not yet succeeded in testing, but my understanding is that the `download` tag should trigger a file download in anything resembling modern (I'm yet unable to understand if `download` and `target="_blank"` will give a flashing window that disappear, an empty window or something else).

jonandernovella replied:
@HenrikeW Will this be addressed here or in the other PR?

pontus commented on app/herdbook.py:
```diff
@@ -505,6 +509,314 @@ def testbreed():
+        present = check_certificate_s3(ind_number=ind["number"])
+        if certificate_exists and present:
+            data = get_certificate_data(ind, user_id)
+            data.update(**form)
```
I still don't like this and don't think it does what's intended (at least not always).

jonandernovella replied:
I still don't understand what you want either. Can you please explain?

pontus replied:
Hrrm, misplaced this comment, should be on the `ind.update` line.

For that; what it looks like it should do is start creating an update query (http://docs.peewee-orm.com/en/latest/peewee/api.html#Model.update which doesn't change anything). Then call the `update_individual` to do update (which is intended to update from a form which is not exactly what we're doing here).

So again, I don't understand how the ind.update can change anything (given as for my current understanding, it just creates an update query that is not executed but immediately forgotten). 

And if it does work as intended, I still don't like it on account on being confusing.

I'd much prefer having a whitelist of allowed changes here (keys in dict `form`) and passing to a new function in `data_access` to update individual according to that form. That could either do a

```
    ind = get_individual(i_number, user_id)
    ind.update(**form).where(Individual.number == i_number).execute()
```   
or
```
    ind = get_individual(i_number, user_id)
    for k,v in form:
        ind.setattr(k,v)
    ind.save()
```   

(both inside `with DATABASE.atomic():`)

pontus replied:
See https://github.com/NBISweden/theherdbook/commit/ef1bfbe4aa910ba6e6d7fe4ac650fda8db325ab8  for some minumum changes (only naming and comments) to lessen confusion. I'd prefer other changes but that would bloat this PR even more, so it's something for the future.

pontus commented on app/herdbook.py:
```diff
@@ -505,6 +509,314 @@ def testbreed():
+            ind.update(**form)
+            da.update_individual(ind, session.get("user_id", None))
+    except Exception as ex:  # pylint: disable=broad-except
+        APP.logger.info("Unexpected error while updating certificate " + str(ex))
```
I think the "can't update paper certificate" case deserves enough treatment that the frontend can give a sensible error.

jonandernovella replied:
The HTTP status code(404) returned by the exception handler is mostly for when calling the API with an individual that does not have a digital certificate uploaded **or the backend is broken**. Hopefully there is a paper one but this is not guaranteed I think this will mostly happen when the user tries to update the certificate of **the wrong individual,** though. I am fine with changing this but note it will not always convey the correct information.

pontus replied:
Ok, after some thinking I believe we really should improve the frontend anyway and this won't be such a big thing.

HenrikeW committed:
add noopener/noreferrer to download links for cert

to make GitHub Code Scanning happier

pontus commented on app/herdbook.py:
```diff
@@ -505,6 +509,314 @@ def testbreed():
+    form = request.json
+    cert_number = ind["digital_certificate"]
+
+    ind.update(**form, certificate=cert_number)
```
Same for database update here as before.

HenrikeW committed:
remove target=blank from download link

following Pontus' comment on thge PR

pontus commented on app/herdbook.py:
```diff
@@ -505,6 +509,314 @@ def testbreed():
+    )
+
+    qr_code = certs.QRHandler(
+        link=settings.service.host + "/individual/" + data["number"] + "/verify",
```
I'm fine with merging this for now, but think the host should really be picked up from the request (likely requires adding 
`proxy_set_header Host $host;` in nginx).

pontus commented on frontend/src/individual_view.tsx:
```diff
@@ -208,9 +208,23 @@ export function IndividualView({ id }: { id: string }) {
+                  }
                 >
-                  Beställ certifikat
+                  Skapa nytt certifikat
+                </Button>
+              )}
+              {user?.canEdit(id) && (
+                <Button
+                  className={style.editButton}
+                  variant="contained"
+                  color="primary"
+                  onClick={() =>
+                    popup(<IndividualCertificate id={id} action={"update"} />)
+                  }
+                >
+                  Uppdatera certifikat
```
After some playing and thinking, I believe we need to do this nicer, even for a MVP.

If there is a paper certificate these should not be shown at all as far as I understand (digital certificate can neither be created nor updated).

If there is a digital certificate "Uppdatera certifikat" should be available.

If there is no certificate, "Skapa nytt certifikat" should be shown.

If I understand the current design correctly, this should be possibly by testing if integer conversion of `certificate` equals integer conversion of `digital_certificate`.

jonandernovella replied:
In the design stage we contemplated these options but later decided not to do this to simplify things.

pontus committed:
Minimum changes to lessen confusion for updates

HenrikeW committed:
show cert buttons depending on cert number

if the rabbit has no cert, you can issue a new (digital) one.
if it has a digital one, you can update and verify it.

HenrikeW committed:
actually show buttons depending on cert number

correcting mistakes from the previous commit
now three cases: no cert at all, a paper cert, a digital cert

jonandernovella committed:
Merge pull request #297 from NBISweden/dev/certifikater_mini_change

Minimum changes to lessen confusion for updates

jonandernovella committed:
Improve content of verify's failure response

jonandernovella committed:
Make black happy

HenrikeW committed:
Fixed user messages during verification process

HenrikeW committed:
error message for file upload error

following a comment from Pontus

jonandernovella committed:
Merge pull request #298 from NBISweden/frontend/verification

frontend for verification process
